src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,635 @@
 
+#![allow(clippy::too_many_arguments)]
+
+use hashbrown::HashMap;
 use num_complex::{Complex64, ComplexFloat};
+use std::cmp::Ordering;
+use std::f64::consts::PI;
+
+use pyo3::exceptions::{PyIndexError, PyTypeError};
 use pyo3::prelude::*;
+use pyo3::types::PySlice;
 use pyo3::wrap_pyfunction;
 use pyo3::Python;
-use std::f64::consts::PI;
 
 use ndarray::prelude::*;
 use numpy::PyReadonlyArray2;
 
+const DEFAULT_ATOL: f64 = 1e-12;
+
+#[derive(FromPyObject)]
+enum SliceOrInt<'a> {
+    Slice(&'a PySlice),
+    Int(isize),
+}
+
+#[pyclass]
+pub struct OneQubitGateErrorMap {
+    error_map: Vec<HashMap<String, f64>>,
+}
+
+#[pymethods]
+impl OneQubitGateErrorMap {
+    #[new]
+    fn new(num_qubits: Option<usize>) -> Self {
+        OneQubitGateErrorMap {
+            error_map: match num_qubits {
+                Some(n) => Vec::with_capacity(n),
+                None => Vec::new(),
+            },
+        }
+    }
+    fn add_qubit(&mut self, error_map: HashMap<String, f64>) {
+        self.error_map.push(error_map);
+    }
+}
+
+#[pyclass(sequence)]
+pub struct OneQubitGateSequence {
+    gates: Vec<(String, Vec<f64>)>,
+    #[pyo3(get)]
+    global_phase: f64,
+}
+
+#[pymethods]
+impl OneQubitGateSequence {
+    #[new]
+    fn new() -> Self {
+        OneQubitGateSequence {
+            gates: Vec::new(),
+            global_phase: 0.,
+        }
+    }
+    fn __getstate__(&self) -> (Vec<(String, Vec<f64>)>, f64) {
+        (self.gates.clone(), self.global_phase)
+    }
+
+    fn __setstate__(&mut self, state: (Vec<(String, Vec<f64>)>, f64)) {
+        self.gates = state.0;
+        self.global_phase = state.1;
+    }
+
+    fn __len__(&self) -> PyResult<usize> {
+        Ok(self.gates.len())
+    }
+
+    fn __getitem__(&self, py: Python, idx: SliceOrInt) -> PyResult<PyObject> {
+        match idx {
+            SliceOrInt::Slice(slc) => {
+                let len = self.gates.len().try_into().unwrap();
+                let indices = slc.indices(len)?;
+                let mut out_vec: Vec<(String, Vec<f64>)> = Vec::new();
+                // Start and stop will always be positive the slice api converts
+                // negatives to the index for example:
+                // list(range(5))[-1:-3:-1]
+                // will return start=4, stop=2, and step=-1
+                let mut pos: isize = indices.start;
+                let mut cond = if indices.step < 0 {
+                    pos > indices.stop
+                } else {
+                    pos < indices.stop
+                };
+                while cond {
+                    if pos < len as isize {
+                        out_vec.push(self.gates[pos as usize].clone());
+                    }
+                    pos += indices.step;
+                    if indices.step < 0 {
+                        cond = pos > indices.stop;
+                    } else {
+                        cond = pos < indices.stop;
+                    }
+                }
```

Comment:
This is a bit repetitive - maybe something like
```rust
let mut pos: isize = indices.start;
let cond = if indices.step < 0 {
    |pos| pos > indices.stop  // `move |pos|` if needed, but I think indices.stop is Copy anyway?
} else {
    |pos| pos < indices.stop
};
while cond(pos) {
    // ...
    pos += indices.step;
}
```
avoids the manual repetition in the `cond` check?

Reply:
I gave this a try locally and it's a bit more involved than I thought because you can't have two arms on an assignment like this as closures because they're all inherently different types. I tried to define it as a dyn function type instead but it wasn't as straightforward as I thought. I'm sure I could play with it a bit more and figure out a pattern that worked, but it didn't seem critical.

Reply:
Oh yeah, good point.  We could put the branch inside the closure?  If it's too much effort, it's not worth bothering, though.

Reply:
Oh I didn't try that, I was fixated on trying to use different closures. :) But, yeah it would work to branch in the closure too. But I'd rather let it merge as is, we can do this in a follow up pretty easily.